File: backend/src/main/java/com/rishan/guardianstack/tariff/motor/MotorTariffService.java
package com.rishan.guardianstack.tariff.motor;

import com.rishan.guardianstack.core.exception.DuplicateResourceException;
import com.rishan.guardianstack.core.exception.ResourceNotFoundException;
import com.rishan.guardianstack.core.response.PaginatedResponse;
import com.rishan.guardianstack.tariff.motor.dto.MotorTariffDTO;
import com.rishan.guardianstack.tariff.motor.dto.MotorTariffFullDTO;
import com.rishan.guardianstack.tariff.motor.dto.MotorTariffMasterAdminViewSearchCriteria;
import com.rishan.guardianstack.tariff.motor.dto.MotorTariffShortView;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.transaction.annotation.Transactional;


@Service
@RequiredArgsConstructor
public class MotorTariffService {

    private final MotorTariffRepository motorTariffRepository;
    private final MotorTariffMapper mapper;

    public PaginatedResponse<MotorTariffShortView> getMotorTariffsForAdminTable(
            MotorTariffMasterAdminViewSearchCriteria searchCriteria) {

        Pageable pageable = PageRequest.of(
                searchCriteria.page(),
                searchCriteria.size(),
                createSort(searchCriteria.sortBy(), searchCriteria.sortDirection()));

        Specification<MotorTariff> spec = MotorTariffMasterAdminViewSpecification
                .withFilters(searchCriteria);

        Page<MotorTariff> motorTariffPage = motorTariffRepository.findAll(spec, pageable);

        List<MotorTariffShortView> shortViews = motorTariffPage.getContent().stream().map(mapper::toMotorTariffShortView).toList();

        return PaginatedResponse.of(
                shortViews,
                motorTariffPage.getNumber(),
                motorTariffPage.getSize(),
                motorTariffPage.getTotalElements(),
                motorTariffPage.getTotalPages(),
                motorTariffPage.hasNext(),
                motorTariffPage.hasPrevious(),
                searchCriteria.sortBy(),
                searchCriteria.sortDirection(),
                String.format("Found %d motorTariffs", motorTariffPage.getTotalElements()));
    }

    // Get an motorTariff by tariffKey
    public MotorTariffDTO getMotorTariffById(Integer tariffKey) {
        MotorTariff motorTariff = motorTariffRepository.findById(tariffKey)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "MotorTariff not found with ID: " + tariffKey));

        return mapper.toMotorTariffDTO(motorTariff);
    }

    public MotorTariffFullDTO getFullMotorTariffById(Integer tariffKey) {
        MotorTariff motorTariff = motorTariffRepository.findById(tariffKey)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "MotorTariff not found with ID: " + tariffKey));
        return mapper.toMotorTariffFullDTO(motorTariff);
    }

    /**
     * Create a new Motor Tariff.
     * Checks for uniqueness against the composite key (Type, Group, Vehicle, Category).
     */
    @Transactional
    public MotorTariffDTO createMotorTariff(MotorTariffDTO dto) {

        motorTariffRepository.findByTariffTypeAndGroupOfVehicleAndTypeOfVehicleAndCategory(
                dto.tariffType(),
                dto.groupOfVehicle(),
                dto.typeOfVehicle(),
                dto.category()
        ).ifPresent(existing -> {
            throw new DuplicateResourceException("typeOfVehicle",
                    "Motor Tariff already exists with ID: " + existing.getTariffKey());
        });

        MotorTariff motorTariff = mapper.toMotorTariff(dto);
        MotorTariff savedTariff = motorTariffRepository.save(motorTariff);

        return mapper.toMotorTariffDTO(savedTariff);
    }

    /**
     * Update an existing Motor Tariff.
     * Ensures that the updated combination doesn't clash with another record.
     */
    @Transactional
    public MotorTariffDTO updateMotorTariff(Integer tariffKey, MotorTariffDTO dto) {

        MotorTariff existingTariff = motorTariffRepository.findById(tariffKey)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "MotorTariff not found with ID: " + tariffKey));

        motorTariffRepository.findByTariffTypeAndGroupOfVehicleAndTypeOfVehicleAndCategory(
                dto.tariffType(),
                dto.groupOfVehicle(),
                dto.typeOfVehicle(),
                dto.category()
        ).ifPresent(matched -> {
            if (!matched.getTariffKey().equals(tariffKey)) {
                throw new DuplicateResourceException("typeOfVehicle",
                        "This vehicle/category combination is already assigned to Tariff ID: " + matched.getTariffKey());
            }
        });

        mapper.toUpdatedMotorTariff(existingTariff, dto);
        MotorTariff updatedTariff = motorTariffRepository.save(existingTariff);
        return mapper.toMotorTariffDTO(updatedTariff);
    }

    // Delete a motor tariff by tariffKey
    public void deleteMotorTariff(Integer tariffKey) {
        if (!motorTariffRepository.existsById(tariffKey)) {
            throw new ResourceNotFoundException("MotorTariff not found with ID: " + tariffKey);
        }
        motorTariffRepository.deleteById(tariffKey);
    }

    public List<String> getHierarchy(String level, String tariffType, String groupOfVehicle, String typeOfVehicle) {
        MotorHierarchyLevel hLevel = MotorHierarchyLevel.fromString(level);

        // fallback
        // to tons
        // if no CC

        return switch (hLevel) {
            case TARIFF_TYPE -> motorTariffRepository.findDistinctTariffTypes();
            case GROUP_OF_VEHICLE -> motorTariffRepository.findDistinctGroupOfVehicles(tariffType);
            case TYPE_OF_VEHICLE -> motorTariffRepository.findDistinctTypeOfVehicle(tariffType, groupOfVehicle);
            case CATEGORY -> {
                List<String> categories = motorTariffRepository.findDistinctCategories(
                        tariffType, groupOfVehicle, typeOfVehicle);

                categories.sort(
                        Comparator.comparingDouble((String s) -> {
                                    double cc = extractCC(s);
                                    return cc == Double.MAX_VALUE ? extractTons(s) : cc; // fallback
                                    // to tons
                                    // if no CC
                                })
                                .thenComparing(s -> isUpto(s) ? 0 : 1));

                yield categories;
            }


            default -> throw new IllegalArgumentException("Invalid level: " + level);
        };
    }

    // Helper method to create Sort object

    /**
     * Create Sort object from sort parameters with validation
     */
    private Sort createSort(String sortBy, String sortDirection) {
        // Allowed sort fields for admin table
        Set<String> allowedTableSortFields = Set.of("tariffKey", "tariffType", "groupOfVehicle",
                "typeOfVehicle",
                "category");

        String validatedSortBy = allowedTableSortFields.contains(sortBy) ? sortBy : "tariffKey";

        Sort.Direction direction = "desc".equalsIgnoreCase(sortDirection)
                ? Sort.Direction.DESC
                : Sort.Direction.ASC;

        return Sort.by(direction, validatedSortBy);
    }

    private double extractCC(String str) {
        Matcher m = Pattern.compile("(\\d+(?:\\.\\d+)?)(?=\\s*CC)").matcher(str);
        if (m.find()) {
            return Double.parseDouble(m.group(1));
        }
        return Double.MAX_VALUE; // no CC found
    }

    private double extractTons(String str) {
        Matcher m = Pattern.compile("(\\d+(?:\\.\\d+)?)(?=\\s*ton)", Pattern.CASE_INSENSITIVE).matcher(str);
        if (m.find()) {
            return Double.parseDouble(m.group(1));
        }
        return Double.MAX_VALUE; // no tons found
    }

    private boolean isUpto(String str) {
        return str.toLowerCase().contains("upto");
    }

}

File: frontend/features/masteradmin/tariff/motor/motor.tariff.service.ts
// features/masteradmin/tariff/motor/motor.tariff.service.ts
import { api } from "@/lib/api.client";
import type { ApiResponse } from "@/types/api.types";
import type {
  MotorTariffDTO,
  MotorTariffFullDTO,
  MotorTariffShortView,
  MotorTariffSearchCriteria,
  MotorHierarchyParams,
} from "./motor.tariff.types";

// ── Master Admin endpoints (/master-admin/tariff/motor) ───────────────────────

/**
 * GET /master-admin/tariff/motor
 * Paginated lightweight table view — MotorTariffShortView
 */
export function getMotorTariffs(
  criteria?: MotorTariffSearchCriteria
): Promise<ApiResponse<MotorTariffShortView[]>> {
  return api.client.get<MotorTariffShortView[]>(
    "/master-admin/tariffs/motor",
    criteria as Record<string, string | number | boolean | null | undefined>
  );
}

/**
 * GET /master-admin/tariff/motor/:tariffKey
 * Full detail view — MotorTariffFullDTO
 */
export function getMotorTariffById(
  tariffKey: number
): Promise<ApiResponse<MotorTariffFullDTO>> {
  return api.client.get<MotorTariffFullDTO>(
    `/master-admin/tariffs/motor/${tariffKey}`
  );
}

/**
 * POST /master-admin/tariff/motor
 * Create a new tariff entry
 */
export function createMotorTariff(
  dto: MotorTariffDTO
): Promise<ApiResponse<MotorTariffDTO>> {
  return api.client.post<MotorTariffDTO>("/master-admin/tariffs/motor", dto);
}

/**
 * PUT /master-admin/tariff/motor/:tariffKey
 * Full update
 */
export function updateMotorTariff(
  tariffKey: number,
  dto: MotorTariffDTO
): Promise<ApiResponse<MotorTariffDTO>> {
  return api.client.put<MotorTariffDTO>(
    `/master-admin/tariffs/motor/${tariffKey}`,
    dto
  );
}

/**
 * DELETE /master-admin/tariff/motor/:tariffKey
 */
export function deleteMotorTariff(
  tariffKey: number
): Promise<ApiResponse<void>> {
  return api.client.delete<void>(`/master-admin/tariffs/motor/${tariffKey}`);
}

// ── Public hierarchy endpoint (/tariff/motor/hierarchy) ───────────────────────

/**
 * GET /tariff/motor/hierarchy
 * Cascading dropdown data for tariff type → group → type → category
 */
export function getMotorHierarchy(
  params: MotorHierarchyParams
): Promise<ApiResponse<string[]>> {
  const queryParams: Record<string, string> = { level: params.level };
  if (params.tariffType) queryParams.tariffType = params.tariffType;
  if (params.groupOfVehicle) queryParams.groupOfVehicle = params.groupOfVehicle;
  if (params.typeOfVehicle) queryParams.typeOfVehicle = params.typeOfVehicle;

  return api.client.get<string[]>("/master-admin/tariffs/motor/hierarchy", queryParams);
}